xonsh/tools.py

```diff
@@ -941,7 +943,10 @@ def ensure_slice(x):
         else:
             raise ValueError('cannot convert {!r} to slice'.format(x))
     except TypeError:
-        raise TypeError('ensure_slice() argument must be a string or a number not {}'.format(type(x)))
+        try:
+            s = slice(*(int(i) for i in x))
+        except (TypeError, ValueError):
+            raise ValueError('cannot convert {!r} to slice'.format(x))
```

Comment:
TypeError is being raised on `int()` not accepting lists and tuple etc. In the traceback you can see the chain of exceptions. This way it's easier for the caller to just except ValueError too.